### PR TITLE
Features/mean var merge cleanup

### PR DESCRIPTION
**Pull request by coquelin77**
## Description

This is a cleanup of the mean and var functions. The moment merging function is rewritten to be used by both mean and var (it is also possible to expand this function to be used for higher order moments).
The var function also supports multi-dimensional axis arguments (the algorithm is the same as it is for mean)

Fixes: #156 

Changes proposed:
- combination of moment merging functions into one function
- add multi-dimensional axis support to var

## Type of change

Select relevant options.

- New feature (non-breaking change which adds functionality)
- Documentation update

Are all split configurations tested and accounted for?
- [x] yes
- [ ] no

Does this change require a documentation update outside of the changes proposed?
- [ ] yes
- [x] no

Does this change modify the behaviour of other functions?
- [ ] yes
- [x] no

Are there code practices which require justification?
- [ ] yes
- [x] no

### CODE REVIEW

File: heat/core/statistics.py [hunk 1]
```diff
@@ -807,99 +811,68 @@ def reduce_means_elementwise(output_shape_i):
...
+        a tuple of the merged moments
 
     References
     ----------
     [1] J. Bennett, R. Grout, P. Pebay, D. Roe, D. Thompson, Numerically stable, single-pass, parallel statistics
         algorithms, IEEE International Conference on Cluster Computing and Workshops, 2009, Oct 2009, New Orleans, LA,
         USA.
     """
+    n1, n2 = m1[-1], m2[-1]
+    mu1, mu2 = m1[-2], m2[-2]
```
Comment (mtar) [inline heat/core/statistics.py]: Check the lengths of m1 and m2?

File: heat/core/statistics.py [hunk 2]
```diff
@@ -1328,32 +1293,54 @@ def reduce_vars_elementwise(output_shape_i):
...
+                    (var_tot[i, 0], var_tot[i, 1], var_tot[i, 2]),
+                    bessel=bessel,
                 )
-                var_tot[0, 0] = merged[0]
-                var_tot[0, 1] = merged[1]
-                var_tot[0, 2] = merged[2]
-
             return var_tot[0][0]
 
     else:  # axis is given
```
Comment (mtar) [inline heat/core/statistics.py]: Small inconsistency: When axis is set, integer tensors are not supported any longer. Numpy does support ints and floats, while pytorch only supports floats in general. This behaviour is also in 'mean'.

File: heat/core/statistics.py [hunk 3]
```diff
@@ -1328,32 +1293,54 @@ def reduce_vars_elementwise(output_shape_i):
...
         # case for var in one dimension
         output_shape = list(x.shape)
-        if isinstance(axis, int):
```
Comment (mtar) [inline heat/core/statistics.py]: Duplicates in axis list: Numpy checks for it, pytorch (and HeAT) doesn't. This is also in 'mean'.

File: heat/core/statistics.py (hunk 2 above)
Comment (coquelin77) [reply heat/core/statistics.py]: because torch does not support ints so we dont either. if its needed, the easiest way to do this would be to cast the input to a float. I dont think that we need to do this though. i dont think that we need to be so religious to numpy

File: heat/core/statistics.py (hunk 3 above)
Comment (coquelin77) [reply heat/core/statistics.py]: does it raise an exception or does it just ignore the duplicates?

**Comment by mtar** [comment]
Would you be so kind to remove the failing test? As of today, torch.logical_xor allows non-bool tensors in the new pytorch version.

**Comment by coquelin77** [comment]
> Would you be so kind to remove the failing test? As of today, torch.logical_xor allows non-bool tensors in the new pytorch version.

yes this is an issue that needs to be addressed. are you working on this? Bjoern wants to put a release out today. I am in a meeting at the moment, but once this is done I will get to work on a fix for the latest version of pytorch

**Comment by coquelin77** [comment]
> Would you be so kind to remove the failing test? As of today, torch.logical_xor allows non-bool tensors in the new pytorch version.

should the fix be to use the bitwise xor?

**Comment by mtar** [comment]
I will make a pull request

**Comment by mtar** [comment]
One last thing, I was playing with var and realized that heat uses Bassel's correction as the default, similar to  [pytorch](https://pytorch.org/docs/stable/torch.html?highlight=var#torch.var). [numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.var.html),however, uses the biased one as default instead. Shall we keep it that way?

**Comment by coquelin77** [comment]
> One last thing, I was playing with var and realized that heat uses Bassel's correction as the default, similar to [pytorch](https://pytorch.org/docs/stable/torch.html?highlight=var#torch.var). [numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.var.html),however, uses the biased one as default instead. Shall we keep it that way?

i think that Bessel's correction should be the default. But this is mostly personal preference. I think that for general statistical methods (as well as more advanced methods) it will provide more statistically reliable results
